src/libexpr/eval.cc

```diff
@@ -2311,6 +2331,8 @@ BackedStringView EvalState::coerceToString(
               v._path.path
             : copyToStore
             ? store->printStorePath(copyPathToStore(context, v.path()))
+            : v.path().accessor->toStringReturnsStorePath()
+            ? store->printStorePath(copyPathToStore(context, SourcePath(v.path().accessor, CanonPath::root))) + v.path().path.absOrEmpty()
```

Comment:
This is new, in order for `readFile "${toString ./.}/.."` to work, just as it did before.
Would not recommend to write that, but similar usages of paths exist in the wild.